From f6a80248e3a5ec61af178314e714ad47fa814639 Mon Sep 17 00:00:00 2001
From: Simon Xue <xxm@rock-chips.com>
Date: Wed, 16 Oct 2019 14:30:25 +0800
Subject: [PATCH] phy: rockchip-inno-combphy: support low power consumption
 mode

Support low power consumption mode when PCIe suspend.

1. Add missing Kconfig and Makefile
2. Add usb3phy_grf_p reset
3. Turn off PCIe phy pd
4. Write some more MAGIC registers to enter low power mode like following:
	/* enter PCIe phy low power mode */
	writel(0x36, priv->mmio + 0x2150);
	writel(0x36, priv->mmio + 0x3150);
	writel(0x02, priv->mmio + 0x21e8);
	writel(0x02, priv->mmio + 0x31e8);
	writel(0x0c, priv->mmio + 0x2080);
	writel(0x0c, priv->mmio + 0x3080);
	writel(0x08, priv->mmio + 0x20c0);
	writel(0x08, priv->mmio + 0x30c0);
	writel(0x08, priv->mmio + 0x2058);

	writel(0x10, priv->mmio + 0x2044);
	writel(0x10, priv->mmio + 0x21a8);
	writel(0x10, priv->mmio + 0x31a8);
	writel(0x08, priv->mmio + 0x2058);
	writel(0x08, priv->mmio + 0x3058);
	writel(0x40, priv->mmio + 0x205c);
	writel(0x40, priv->mmio + 0x305c);
	writel(0x08, priv->mmio + 0x2184);
	writel(0x08, priv->mmio + 0x3184);
	writel(0x00, priv->mmio + 0x2150);
	writel(0x00, priv->mmio + 0x3150);
	writel(0x10, priv->mmio + 0x20e0);
	writel(0x00, priv->mmio + 0x21e8);
	writel(0x00, priv->mmio + 0x31e8);

Change-Id: I2dc7f99e54d336f45a13b2842bde1a02300311fc
Signed-off-by: Simon Xue <xxm@rock-chips.com>
---
 drivers/phy/rockchip/Kconfig                  |  8 +++
 drivers/phy/rockchip/Makefile                 |  1 +
 .../phy/rockchip/phy-rockchip-inno-combphy.c  | 67 +++++++++++++++++--
 3 files changed, 69 insertions(+), 7 deletions(-)

diff --git a/drivers/phy/rockchip/Kconfig b/drivers/phy/rockchip/Kconfig
index 3cff56855822..00c99cf7d617 100644
--- a/drivers/phy/rockchip/Kconfig
+++ b/drivers/phy/rockchip/Kconfig
@@ -15,6 +15,14 @@ config PHY_ROCKCHIP_EMMC
 	help
 	  Enable this to support the Rockchip EMMC PHY.
 
+config PHY_ROCKCHIP_INNO_COMBPHY
+	tristate "Rockchip INNO USB 3.0 and PCIe COMBPHY Driver"
+	depends on (ARCH_ROCKCHIP && OF) || COMPILE_TEST
+	select GENERIC_PHY
+	help
+	  Enable this to support the Rockchip SoCs COMBPHY.
+	  If unsure, say N.
+
 config PHY_ROCKCHIP_INNO_HDMI_PHY
 	tristate "Rockchip INNO HDMI PHY Driver"
 	depends on ARCH_ROCKCHIP && OF
diff --git a/drivers/phy/rockchip/Makefile b/drivers/phy/rockchip/Makefile
index b98dfccc90fd..43d4c301a282 100644
--- a/drivers/phy/rockchip/Makefile
+++ b/drivers/phy/rockchip/Makefile
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 obj-$(CONFIG_PHY_ROCKCHIP_DP)		+= phy-rockchip-dp.o
 obj-$(CONFIG_PHY_ROCKCHIP_EMMC)		+= phy-rockchip-emmc.o
+obj-$(CONFIG_PHY_ROCKCHIP_INNO_COMBPHY)	+= phy-rockchip-inno-combphy.o
 obj-$(CONFIG_PHY_ROCKCHIP_INNO_HDMI_PHY)	+= phy-rockchip-inno-hdmi-phy.o
 obj-$(CONFIG_PHY_ROCKCHIP_INNO_USB2)	+= phy-rockchip-inno-usb2.o
 obj-$(CONFIG_PHY_ROCKCHIP_INNO_VIDEO_COMBO_PHY)	+= phy-rockchip-inno-video-combo-phy.o
diff --git a/drivers/phy/rockchip/phy-rockchip-inno-combphy.c b/drivers/phy/rockchip/phy-rockchip-inno-combphy.c
index 1832950ffdbb..df1e052edb51 100644
--- a/drivers/phy/rockchip/phy-rockchip-inno-combphy.c
+++ b/drivers/phy/rockchip/phy-rockchip-inno-combphy.c
@@ -27,7 +27,8 @@ enum rockchip_combphy_rst {
 	PHY_POR_RSTN	= 1,
 	PHY_APB_RSTN	= 2,
 	PHY_PIPE_RSTN	= 3,
-	PHY_RESET_MAX	= 4,
+	PHY_GRF_P_RSTN  = 4,
+	PHY_RESET_MAX	= 5,
 };
 
 struct combphy_reg {
@@ -100,6 +101,8 @@ static const char *get_reset_name(enum rockchip_combphy_rst rst)
 		return "combphy-apb";
 	case PHY_PIPE_RSTN:
 		return "combphy-pipe";
+	case PHY_GRF_P_RSTN:
+		return "usb3phy_grf_p";
 	default:
 		return "invalid";
 	}
@@ -269,6 +272,11 @@ static int phy_pcie_init(struct rockchip_combphy_priv *priv)
 
 	grfcfg = &priv->cfg->grfcfg;
 
+	/* reset PCIe phy to default configuration */
+	reset_control_assert(priv->rsts[PHY_POR_RSTN]);
+	reset_control_assert(priv->rsts[PHY_APB_RSTN]);
+	reset_control_assert(priv->rsts[PHY_PIPE_RSTN]);
+
 	reset_control_deassert(priv->rsts[PHY_POR_RSTN]);
 	/* Wait PHY power on stable */
 	udelay(5);
@@ -456,6 +464,15 @@ static int rockchip_combphy_exit(struct phy *phy)
 	 */
 	clk_disable_unprepare(priv->ref_clk);
 
+	priv->phy_initialized = false;
+
+	/* in case of waiting phy PLL lock timeout */
+	if (priv->phy_type == PHY_TYPE_PCIE) {
+		reset_control_assert(priv->rsts[PHY_GRF_P_RSTN]);
+		udelay(5);
+		reset_control_deassert(priv->rsts[PHY_GRF_P_RSTN]);
+	}
+
 	return 0;
 }
 
@@ -511,7 +528,8 @@ static int rockchip_combphy_power_off(struct phy *phy)
 
 	grfcfg = &priv->cfg->grfcfg;
 
-	if (priv->phy_type == PHY_TYPE_USB3) {
+	if (priv->phy_type == PHY_TYPE_USB3 ||
+	    priv->phy_type == PHY_TYPE_PCIE) {
 		/*
 		 * Check if lane 0 powerdown is already
 		 * controlled by grf and in P3 state.
@@ -784,6 +802,10 @@ static int rk1808_combphy_cfg(struct rockchip_combphy_priv *priv)
 	}
 
 	if (priv->phy_type == PHY_TYPE_PCIE) {
+		/* turn on pcie phy pd */
+		writel(0x08400000, priv->mmio + 0x0);
+		writel(0x03030000, priv->mmio + 0x8);
+
 		/* Adjust Lane 0 Rx interface timing */
 		writel(0x20, priv->mmio + 0x20ac);
 		writel(0x12, priv->mmio + 0x20c8);
@@ -807,8 +829,8 @@ static int rk1808_combphy_cfg(struct rockchip_combphy_priv *priv)
 		writel(0xbc, priv->mmio + 0x45d4);
 
 		/* Boost pre-emphasis */
-		writel(0x8b, priv->mmio + 0x21b8);
-		writel(0x8b, priv->mmio + 0x31b8);
+		writel(0xaa, priv->mmio + 0x21b8);
+		writel(0xaa, priv->mmio + 0x31b8);
 	} else if (priv->phy_type == PHY_TYPE_USB3) {
 		/*
 		 * Disable PHY Lane 1 which isn't needed
@@ -889,7 +911,7 @@ static int rk1808_combphy_cfg(struct rockchip_combphy_priv *priv)
 		 * largest swing and "0000" the smallest.
 		 */
 		reg = readl(priv->mmio + 0x21b8);
-		reg = (reg & ~0xf0) | 0xa0;
+		reg = (reg & ~0xf0) | 0xe0;
 		writel(reg, priv->mmio + 0x21b8);
 
 		/*
@@ -913,8 +935,39 @@ static int rk1808_combphy_cfg(struct rockchip_combphy_priv *priv)
 static int rk1808_combphy_low_power_control(struct rockchip_combphy_priv *priv,
 					    bool en)
 {
-	if (priv->phy_type != PHY_TYPE_USB3)
-		return -EINVAL;
+	if (priv->phy_type != PHY_TYPE_USB3) {
+		/* turn off pcie phy pd */
+		writel(0x08400840, priv->mmio + 0x0);
+		writel(0x03030303, priv->mmio + 0x8);
+
+		/* enter PCIe phy low power mode */
+		writel(0x36, priv->mmio + 0x2150);
+		writel(0x36, priv->mmio + 0x3150);
+		writel(0x02, priv->mmio + 0x21e8);
+		writel(0x02, priv->mmio + 0x31e8);
+		writel(0x0c, priv->mmio + 0x2080);
+		writel(0x0c, priv->mmio + 0x3080);
+		writel(0x08, priv->mmio + 0x20c0);
+		writel(0x08, priv->mmio + 0x30c0);
+		writel(0x08, priv->mmio + 0x2058);
+
+		writel(0x10, priv->mmio + 0x2044);
+		writel(0x10, priv->mmio + 0x21a8);
+		writel(0x10, priv->mmio + 0x31a8);
+		writel(0x08, priv->mmio + 0x2058);
+		writel(0x08, priv->mmio + 0x3058);
+		writel(0x40, priv->mmio + 0x205c);
+		writel(0x40, priv->mmio + 0x305c);
+		writel(0x08, priv->mmio + 0x2184);
+		writel(0x08, priv->mmio + 0x3184);
+		writel(0x00, priv->mmio + 0x2150);
+		writel(0x00, priv->mmio + 0x3150);
+		writel(0x10, priv->mmio + 0x20e0);
+		writel(0x00, priv->mmio + 0x21e8);
+		writel(0x00, priv->mmio + 0x31e8);
+
+		return 0;
+	}
 
 	if (en) {
 		/* Lane 0 tx_biasen disable */
-- 
2.35.3

